chia/full_node/tx_processing_queue.py

```diff
@@ -123,25 +132,69 @@ def put(self, tx: TransactionQueueEntry, peer_id: bytes32 | None, high_priority:
                 raise TransactionQueueFull(f"Transaction queue full for peer {peer_id}")
         self._queue_length.release()  # increment semaphore to indicate that we have a new item in the queue
 
+    def _cleanup_peer_queues(self) -> None:
+        """
+        Removes empty peer queues and updates the cursor accordingly.
+        """
+        new_peer_map = []
+        for peer_id in self._index_to_peer_map:
+            if self._queue_dict[peer_id].empty():
+                self._queue_dict.pop(peer_id, None)
+                self._deficit_counters.pop(peer_id, None)
+            else:
+                new_peer_map.append(peer_id)
+        self._index_to_peer_map = new_peer_map
+        if len(self._index_to_peer_map) > 0:
+            self._list_cursor %= len(self._index_to_peer_map)
+        else:
+            self._list_cursor = 0
+
     async def pop(self) -> TransactionQueueEntry:
         await self._queue_length.acquire()
         if not self._high_priority_queue.empty():
             return self._high_priority_queue.get()
-        result: TransactionQueueEntry | None = None
         while True:
-            peer_queue = self._queue_dict[self._index_to_peer_map[self._list_cursor]]
-            if not peer_queue.empty():
-                _, result = peer_queue.get()
-            self._list_cursor += 1
-            if self._list_cursor > len(self._index_to_peer_map) - 1:
-                # reset iterator
-                self._list_cursor = 0
-                new_peer_map = []
-                for peer_id in self._index_to_peer_map:
-                    if self._queue_dict[peer_id].empty():
-                        self._queue_dict.pop(peer_id)
-                    else:
-                        new_peer_map.append(peer_id)
-                self._index_to_peer_map = new_peer_map
-            if result is not None:
-                return result
+            # Map of peer ID to its top transaction's advertised cost
+            top_txs_advertised_costs: dict[bytes32, uint64] = {}
+            # Let's see if a peer can afford to send its top transaction
+            num_peers = len(self._index_to_peer_map)
+            if num_peers == 0:
+                continue
+            start = self._list_cursor
+            for offset in range(num_peers):
+                peer_index = (start + offset) % num_peers
+                peer_id = self._index_to_peer_map[peer_index]
+                peer_queue = self._queue_dict[peer_id]
+                if peer_queue.empty():
+                    continue
+                # There is no peek method so we access the internal `queue`
+                _, entry = peer_queue.queue[0]
+                tx_info = entry.peers_with_tx.get(peer_id)
+                if tx_info is None:
+                    top_tx_advertised_cost = max(
+                        (t.advertised_cost for t in entry.peers_with_tx.values()), default=self._max_tx_clvm_cost
```

Comment:
in what case is this default value relevant? would you ever get here if `entry.peers_with_tx` is empty?